### ISSUE TITLE
Silent Issue with sonarlint - PACKAGE-PROTECTED

### ISSUE THREAD

**Issue by clement-brodu** [Riverside-Software/sonar-openedge#882]
Hello,

I have a issue when I execute Sonarlint with the files that are in the zip.
[Src.zip](https://github.com/Riverside-Software/sonar-openedge/files/6421357/Src.zip)

In my case I'm using PACKAGE-PROTECTED visibility for Property with a include.

Sometimes it works, but sometimes not. 
For example, if it works the first time, and then I add a new line anywhere, then it doesn't work.

if it does not work, I have no error even though I have unused variables.

Version :
- sonarlint 5.5.1509 
- cabl : 2.12.1
- openedge 12.2.4

### May be it could help :  
I made some tests and I saw that eu.rssw.openedge.rcode.rcode-reader plugin can't parse my rcode file (but may be it's another problem).
It seems that PACKAGE-PROTECTED is not implemented in rcode-reader.
Here is the stack trace with rcode-reader 2.13.0 :
```
java.lang.IndexOutOfBoundsException
[java.nio.ByteBuffer.wrap(ByteBuffer.java:375), eu.rssw.pct.RCodeInfo.readNullTerminatedString(RCodeInfo.java:367), 
eu.rssw.pct.RCodeInfo.readNullTerminatedString(RCodeInfo.java:358), 
eu.rssw.pct.elements.v12.PropertyElementV12.fromDebugSegment(PropertyElementV12.java:47), 
eu.rssw.pct.elements.v12.TypeInfoV12.newTypeInfo(TypeInfoV12.java:125), 
eu.rssw.pct.RCodeInfo.processTypeBlock(RCodeInfo.java:271), eu.rssw.pct.RCodeInfo.<init>(RCodeInfo.java:144), 
eu.rssw.pct.RCodeInfo.<init>(RCodeInfo.java:104), sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method), 
sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62), 
sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45), 
java.lang.reflect.Constructor.newInstance(Constructor.java:423), 
org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:83), 
org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105), 
org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:60), 
org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:235), 
org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:247), 
...
```

[Obj.zip](https://github.com/Riverside-Software/sonar-openedge/files/6421358/Obj.zip)

**Comment by gquerret** [comment]
Will definitely have a look, I may have to contact you offline.

**Comment by gquerret** [comment]
Issue confirmed, fix will probably go into a hotfix.

**Comment by gquerret** [comment]
@clement-brodu Can you try this version ?
https://ci.rssw.eu/job/sonar-openedge/job/hotfix%252FRedQualityGate/2/

Though it's a development snapshot, I'd consider it safe to use on your production SQ instance. But if you want to stay on the super-safe side, you can try it on a test SQ instance.

**Comment by clement-brodu** [comment]
For rcode parsing, it seems to be better, I will make additionnal tests.

For sonarlint, I still have some problems :

### Case 1
```openedge
BLOCK-LEVEL ON ERROR UNDO, THROW.

CLASS Core.BugSonar01: 
    
    {Core/PackageProtected.i}
        
    DEFINE VARIABLE vtoto AS INTEGER NO-UNDO.
    
    DEFINE PACKAGE-PROTECTED PROPERTY Object0 AS Progress.Lang.Object NO-UNDO 
    GET.
    SET.
    
    DEFINE PUBLIC PROPERTY Object1 AS Progress.Lang.Object NO-UNDO 
    GET.
    SET.
    
    METHOD PUBLIC VOID HelloWorld(  ):
        DEFINE VARIABLE vBug AS CHARACTER NO-UNDO.
        MESSAGE  "Hello"  VIEW-AS ALERT-BOX INFORMATION BUTTONS OK IN WINDOW CURRENT-WINDOW.
        RETURN.
    END METHOD.

END CLASS.
```
log:
```
Parsing [uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls]
Parsing hierarchy of '[uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls]' - Relative 'Core\BugSonar01.cls'
  RCode found: 'D:\Openedge\workspace-oe12\DemoSonar\Obj\Core\BugSonar01.r'
Parsing rcode D:\Openedge\workspace-oe12\DemoSonar\Obj\Core\BugSonar01.r
Injecting type info 'Type info Core.BugSonar01 - Parent Progress.Lang.Object'
Listing file for '[uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls]' not found or contains space character - Was looking for 'null'
15 milliseconds to generate ParseUnit
ActiveRule - Internal key  - Repository rssw-oe-main - Rule eu.rssw.antlr.proparse.checks.GlobalDefine
Error during rule execution for [uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls]
java.lang.UnsupportedOperationException: hasRelativePath
	at org.sonarsource.sonarlint.core.container.analysis.filesystem.DefaultFilePredicates.hasRelativePath(DefaultFilePredicates.java:67)
	at eu.rssw.antlr.proparse.checks.GlobalDefine.execute(Unknown Source)
	at org.sonar.plugins.openedge.api.checks.OpenEdgeProparseCheck.sensorExecute(OpenEdgeProparseCheck.java:57)
	at org.sonar.plugins.openedge.sensor.OpenEdgeProparseSensor.parseMainFile(OpenEdgeProparseSensor.java:425)
	at org.sonar.plugins.openedge.sensor.OpenEdgeProparseSensor.execute(OpenEdgeProparseSensor.java:194)
	at org.sonarsource.sonarlint.core.analyzer.sensor.SensorsExecutor.executeSensor(SensorsExecutor.java:80)
	at org.sonarsource.sonarlint.core.analyzer.sensor.SensorsExecutor.execute(SensorsExecutor.java:71)
	at org.sonarsource.sonarlint.core.container.analysis.AnalysisContainer.doAfterStart(AnalysisContainer.java:132)
	at org.sonarsource.sonarlint.core.container.ComponentContainer.startComponents(ComponentContainer.java:126)
	at org.sonarsource.sonarlint.core.container.ComponentContainer.execute(ComponentContainer.java:111)
	at org.sonarsource.sonarlint.core.container.storage.StorageAnalyzer.analyze(StorageAnalyzer.java:75)
	at org.sonarsource.sonarlint.core.container.storage.StorageContainerHandler.analyze(StorageContainerHandler.java:85)
	at org.sonarsource.sonarlint.core.ConnectedSonarLintEngineImpl.lambda$analyze$0(ConnectedSonarLintEngineImpl.java:152)
	at org.sonarsource.sonarlint.core.ConnectedSonarLintEngineImpl.withReadLock(ConnectedSonarLintEngineImpl.java:349)
	at org.sonarsource.sonarlint.core.ConnectedSonarLintEngineImpl.withReadLock(ConnectedSonarLintEngineImpl.java:339)
	at org.sonarsource.sonarlint.core.ConnectedSonarLintEngineImpl.analyze(ConnectedSonarLintEngineImpl.java:149)
	at org.sonarlint.eclipse.core.internal.engine.connected.ConnectedEngineFacade.lambda$13(ConnectedEngineFacade.java:354)
	at org.sonarlint.eclipse.core.internal.engine.connected.ConnectedEngineFacade.withEngine(ConnectedEngineFacade.java:147)
	at org.sonarlint.eclipse.core.internal.engine.connected.ConnectedEngineFacade.runAnalysis(ConnectedEngineFacade.java:353)
	at org.sonarlint.eclipse.core.internal.jobs.AnalyzeConnectedProjectJob.runAnalysis(AnalyzeConnectedProjectJob.java:70)
	at org.sonarlint.eclipse.core.internal.jobs.AnalyzeConnectedProjectJob.runAnalysis(AnalyzeConnectedProjectJob.java:1)
	at org.sonarlint.eclipse.core.internal.jobs.AbstractAnalyzeProjectJob.run(AbstractAnalyzeProjectJob.java:389)
	at org.sonarlint.eclipse.core.internal.jobs.AbstractAnalyzeProjectJob.runAnalysisAndUpdateMarkers(AbstractAnalyzeProjectJob.java:199)
	at org.sonarlint.eclipse.core.internal.jobs.AbstractAnalyzeProjectJob.doRun(AbstractAnalyzeProjectJob.java:161)
	at org.sonarlint.eclipse.core.internal.jobs.AbstractSonarProjectJob.run(AbstractSonarProjectJob.java:45)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

1 files proparse'd, 0 XML files, 0 listing files, 0 failure(s), 0 NCLOCs
```

### Case 2 

```openedge
BLOCK-LEVEL ON ERROR UNDO, THROW.

CLASS Core.BugSonar01: 
    
    {Core/PackageProtected.i}
        
    DEFINE VARIABLE vtoto AS INTEGER NO-UNDO.
    
    DEFINE PACKAGE-PROTECTED PROPERTY Object0 AS Progress.Lang.Object NO-UNDO 
    GET.
    SET.
    
    DEFINE PUBLIC PROPERTY Object1 AS Progress.Lang.Object NO-UNDO 
    GET.
    PACKAGE-PROTECTED SET.
    
    METHOD PUBLIC VOID HelloWorld(  ):
        DEFINE VARIABLE vBug AS CHARACTER NO-UNDO.
        MESSAGE  "Hello"  VIEW-AS ALERT-BOX INFORMATION BUTTONS OK IN WINDOW CURRENT-WINDOW.
        RETURN.
    END METHOD.

END CLASS.
```

```
Parsing [uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls]
Parsing hierarchy of '[uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls]' - Relative 'Core\BugSonar01.cls'
  RCode found: 'D:\Openedge\workspace-oe12\DemoSonar\Obj\Core\BugSonar01.r'
Parsing rcode D:\Openedge\workspace-oe12\DemoSonar\Obj\Core\BugSonar01.r
Injecting type info 'Type info Core.BugSonar01 - Parent Progress.Lang.Object'
Listing file for '[uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls]' not found or contains space character - Was looking for 'null'
Syntax error -- Src/Core/BugSonar01.cls:15:23 -- no viable alternative at input 'PACKAGE-PROTECTED SET'
Error during code parsing for [uri=file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls] at position Src/Core/BugSonar01.cls:15:23
Initializing metadata of file file:/D:/Openedge/workspace-oe12/DemoSonar/Src/Core/BugSonar01.cls
```

**Comment by gquerret** [comment]
Case 1 is related to issue #868. Code in the OE plugin was fixed, but the GlobalDefine rule is still using the relativePath() method which throws an exception in SonarLint 8. Will be fixed.
Case 2 is being investigated.

**Comment by clement-brodu** [comment]
For rcode-reader, it's OK, no more exception when parsing rcode file 👍

**Comment by gquerret** [comment]
Fix will be included in 2.13.1. Other issues mentioned in comment are moved to #883 and #884
